### ISSUE TITLE
0.3 release checklist

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/numpyro#598]
Here we list issues that we want to address for the next release:

- [x] Port `pyro.infer.reparam` #599
- [x] Remove all `base_param_map`/`transform_with_intermediates` stuffs to simplify our codebase. Remove `AutoContinuousELBO`. #601
- [x] Update `neutra`, `neal_funnel`, `eight_schools` examples to reflect the usage of `reparam`
- [x] Update the pattern to define improper prior #605
- [x] Move `autoguide` to the `infer` module. 
- [x] Refactor `initialization` strategies, following Pyro versions. Revise the pattern of subsitute_fn because it is used incorrectly, as mention at [this comment](https://github.com/pyro-ppl/numpyro/pull/595/files#diff-fee397191e0cc75a2e33eea8f7ecb6d5R335) #606
- [x] Unifies the pattern of returning statistics for deterministic sites
- [x] #585 Addresses compiling issues
- [x] Add some `control_flow` primitives #595
- ~[ ] #567 Enhance `kinetic_energy` API, based on recent refactoring in Pyro. This should allow users to set initial mass matrix #536~ **edit** It is not clear which API we should change... so I would like to defer this to future PRs when structure mass matrix is required.
- [ ] Initial support for models with discrete latent variable. We can defer advanced stuff such as `markov` for later releases.
- [x] Update readme eight school example using reparam.
- [x] Update notebooks
- [x] Update jax/jaxlib to the latest version:
    + Use the convention `import jax.numpy as jnp` #590
    + Use the upstream logsumexp again
    + Use `lax.switch` for multiple branches
    + Use the upstream poisson sampler
- [x] some tweaks
    + remove device put in poisson
    + match Pyro api for `condition`, `substitute`, `scale`, Pareto, MultivariateAffineTransform...
    + verify all the docs are generated as expected
    + PRNGKey log_prob issue
    + remove contrib/distributions

**Comment by fehiepsi** [comment]
@neerajprad I want to remove the support for improper prior through `pyro.param`. The reasons are:
+ We can achieve the same purpose by using `sample(name, dist.Foo().mask(False))`. Using `sample` is better because we can draw samples from the (improper) prior to get initial values; rather than a fixed value from `param` statement.
+ The distinguish between `param` and `sample` is clearer.
+ Keeps the behavior like in Pyro, without losing functionality.

There are some other technical benefits:
+ I want to resolve the issue to get initial values [here](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/util.py#L330). There we used the same `subkey` for all sites. I think an easy fix is to use `substitute(seed(model, ...), substitute_fn=...`, so each `sample` site will have `rng_key` before applying `substitute_fn`. Then we can use that `rng_key` inside init strategies. But this won't work for `param` statement.
+ With that, we can remove `skip_params` flag. This allows users specify `init_strategy` using `init_to_uniform` in addition to `init_to_uniform(radius=3)` (Pyro behavior). I think that the current pattern has made @martinjankowiak confused.

What do you think?

**Comment by neerajprad** [comment]
Removing support seems fine to me and +1 to more uniformity with Pyro's syntax. We should just make sure that we document and raise a warning/error that tells users how to change their code. I'm curious to see your `.mask(False)` solution. Let's discuss that when you have a PR up.

**Comment by fehiepsi** [comment]
@neerajprad I think that my last refactoring job is to split `SA` implementation from `mcmc.py` file, as you requested in the last release. Another point that you wanted to discuss was whether or not to include `deterministic` sites to the output of `AutoGuide.get_posterior`. What is your opinion about it now? (the current behavior is to include `deterministic` sites to the output) I think that's all pending tasks that I can remember. If you want to discuss or address something else for 0.3 release, please feel free to edit the topic tasks.

**Comment by neerajprad** [comment]
> @neerajprad I think that my last refactoring job is to split SA implementation from mcmc.py file, as you requested in the last release. 

This is purely an internal code organization related change, so it shouldn't block us from releasing. 

> Another point that you wanted to discuss was whether or not to include deterministic sites to the output of AutoGuide.get_posterior. What is your opinion about it now? (the current behavior is to include deterministic sites to the output) 

I don't recall the discussion right now, and the arguments for each. Could you point me to it? In any case, I don't think any of these issues should be blocking us from doing a release. 

The only thing that will be nice to have is usage of `scan` in some tutorial / example. Can we modify our time series forecasting example to use your modified `scan`. We can do this after the release btw and update the tutorial.

**Comment by fehiepsi** [comment]
The discussion that we deferred for later about autoguide is from https://github.com/pyro-ppl/numpyro/pull/601#issuecomment-634387357

I think I can modify ts forecasting example to illustrate scan. Nice suggestion!! Thanks ;)

**Comment by fehiepsi** [comment]
I created this issue to track down issues for 0.3 release. Most of them are addressed or in reviewing so I will close this now.
